docs/source/architecture/journal.rst

```diff
@@ -0,0 +1,379 @@
+
+The BlockStore is a persistent on-disk store of all Blocks in the current
+chain. When the validator is started, the contents of the BlockStore is trusted
+to be the current state of the world. The BlockStore also stores the identifier
+of the chain head.
+
+All blocks stored here are formally complete. The BlockStore allows blocks to
+be accessed via Block ID, additionally blocks may be accessed via Batch ID or
+Transaction ID.  I.e. GetBlockByTransactionId, GetBatchByTransactionId and
+GetBlockByBatchId.
+
+The BlockStore maintains internal mappings of Transaction-to-Block and
+Batch-to-Block. These may be rebuilt if missing or corrupt, this rebuild should
+be done during start up, not during the course of normal operation. These
+mappings should be stored in a format that is cached to disk, so they are not
+required to be held in memory at all times. As the block-chain grows these will
+become quite large.
+
+The BlockStore provides an atomic method for updating the current head of the
+chain. In order for the BlockStore to switch forks, it is provided with a list
+of blocks in the new chain to commit, and a list of blocks in the old chain to
+decommit. These lists are the blocks in each fork back to the common root.
+
+The BlockCache
+==============
+
+The Block Cache holds the working set of blocks for the validator and tracks the
+processing state. This processing state is tracked as valid, invalid, or
+unknown. Valid blocks are blocks that have been proven to be valid by the
+ChainController. Invalid blocks are blocks that failed validation or have an
+invalid block as a predecessor. Unknown are blocks that have not yet completed
+validation, usually having just arrived from the Completer.
+
+The BlockCache is an in-memory construct, not persisted. It is rebuilt by demand
+when the system is started. 
+
+If a block is not present in the BlockCache it will look in the BlockStore for
+the block. If it is not found or the lookup fails, the block is unknown to the
+system. If the block is found in the BlockStore it is loaded into the
+BlockCache and marked as valid. All blocks in the BlockStore are considered
+valid.
+
+The BlockCache keeps blocks that are currently relevant, tracked by the last
+time the block was accessed. Periodically, the blocks that have not been
+accessed recently are purged from the block cache. The time horizon for purging
+and the frequency of purging are either configured or determined automatically
+based on the average block time of the system.
+
+The Completer 
+=============
+
+The Completer is responsible for making sure Blocks and Batches are complete
+before they are delivered. Blocks are considered formally complete once all of
+their predecessors have been delivered to the ChainController and their batches
+field contains all the Batches specified in the BlockHeader’s batch_ids list.
+The batches field is also expected be in the same order as the batch_ids. Once
```

Comment:
Missing preposition "to":
The batches field is also expected **to** be in the same order as the batch_ids.